### PR TITLE
load resources accross ajax [new feature]

### PR DESCRIPTION
**Pull request by Ivanov-Anton**
load resources accross ajax
solved: #711 

- contractor/customer/vendor
- account
- gateway/orig gw/term gw

ajax
---------------------------------------------------------------------------

 **has gateway filter**:
- [x] import_dialpeers
- [x] auth_logs (Cdr::AuthLog)
- [x] rtp_statistics (Cdr::RtpStatistic)
- [x] dialpeers (Dialpeer)
- [x] customers_auths (CustomersAuth)
**has account filter**:
- [x] import_dialpeers
- [x] invoices (Billing::Invoice)
- [x] customers_auths (CustomersAuth)
- [x] dialpeers (Dialpeer)
 **has contractor filter**
**vendor**
- [x] import_dialpeers
- [x] import_gateway_groups
- [x] gateway_groups ()
- [x] active_calls (RealtimeData::ActiveCall)
- [x] cdrs (Cdr::Cdr)
- [x] vendor_traffics (Report::VendorTraffic)
- [x] vendor_traffic_schedulers (Report::VendorTrafficScheduler)
- [x] dialpeers (Dialpeer)
- [x] static_routes (Routing::RoutingPlanStaticRoute)
**customer**
- [x] cdrs
- [x] active_calls
- [x] custom_cdrs
- [x] custom_cdr_schedulers
- [x] customer_traffics
- [x] customer_traffic_schedulers
- [x] report_realtime_bad_routings
- [x] report_realtime_origination_performances
- [x] report_realtime_termination_distributions
- [x] customers_auths
- [x] api_accesses
**contractor**
- [x] accounts
- [x] import_gateways
- [x] import_accounts
- [x] billing_contacts 
- [x] invoices 
- [x] gateways

### CODE REVIEW

File: lib/resource_dsl/filter_ajax.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        class_name = options[:class_name]
+      elsif options[:class_name].blank?
+        if name.to_s == 'customer'
+          class_name = 'contractor'.capitalize
+        elsif name.to_s == 'vendor'
+          class_name = 'contractor'.capitalize
+        else
+          class_name = name.to_s.capitalize
+        end
+      end
```
Comment (gigorok) [inline lib/resource_dsl/filter_ajax.rb:16]: if options[:class_name].blank?
   options[:class_name] = name.classify.constantize
end

File: lib/resource_dsl/filter_ajax.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+      collection = proc {
+        resource_id = params.fetch(:q, {})[:"#{name}_id_eq"]
+        resource_id ? Object.const_get(class_name).where(id: resource_id) : []
```
Comment (gigorok) [inline lib/resource_dsl/filter_ajax.rb:20]: options[:class_name].where(id: resource_id)

File: lib/resource_dsl/filter_ajax.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      }
+      options = options.merge(as: :select, collection: collection)
+      options[:input_html] ||= {}
+      options[:input_html][:class] = 'chosen-ajax'
```
Comment (gigorok) [inline lib/resource_dsl/filter_ajax.rb:24]: what if the class was passed through options ?

File: lib/resource_dsl/filter_ajax.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      name = 'contractor' if %w[vendor customer].include?(name)
+      options[:input_html][:'data-path'] = "/#{class_name.downcase}s/find" if options[:input_html].blank?
```
Comment (gigorok) [inline lib/resource_dsl/filter_ajax.rb:26]: options[:input_html][:'data-path'] = "/#{options[:class_name].to_s.underscore.pluralize}/find" if options.dig(:input_html, :'data-path').blank?

File: lib/resource_dsl/filter_ajax.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      end
+
+      collection = proc {
+        resource_id = params.fetch(:q, {})[:"#{name}_id_eq"]
+        resource_id ? Object.const_get(class_name).where(id: resource_id) : []
+      }
+      options = options.merge(as: :select, collection: collection)
+      options[:input_html] ||= {}
+      options[:input_html][:class] = 'chosen-ajax'
+      name = 'contractor' if %w[vendor customer].include?(name)
```
Comment (gigorok) [inline lib/resource_dsl/filter_ajax.rb:25]: remove it

File: lib/resource_dsl/filter_ajax.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      options[:input_html][:'data-path'] = "/#{class_name.downcase}s/find" if options[:input_html].blank?
+
+      filter "#{name}_id_eq", options
```
Comment (gigorok) [inline lib/resource_dsl/filter_ajax.rb:28]: filter "#{options[:class_name].to_s.underscore}_id_eq", options

File: app/admin/billing/account.rb [hunk 7]
```diff
@@ -1,8 +1,9 @@
 # frozen_string_literal: true
 
 ActiveAdmin.register Account do
+  extend ActiveAdmin::FindAction
   menu parent: 'Billing', priority: 10
-
+  find_action
```
Comment (gigorok) [inline app/admin/billing/account.rb:6]: search_support!

File: lib/resource_dsl/search_support.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module ResourceDSL
+  module ActiveSearch
+    def search_support!
+      collection_action :search, method: :get do
+        data = resource_class.ransack(params[:q]).result
+        result = data.map do |item|
+          %W[#{item.id} #{item.display_name}]
```
Comment (gigorok) [inline lib/resource_dsl/search_support.rb:9]: {id: item.id, value: item.display_name}

File: app/assets/javascripts/ajax-chosen.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                    success: function (data) {
+                        if (data.length == 0) {
+                            return true;
+                        }
+                        // Clear options
+                        select.find('option:not(:selected)').remove();
+                        chosen.find('ul.chosen-results').html('');
+                        // Set new options
+                        $.each(data, function (i, item) {
+                            select.append('<option value="' + item[0] + '">' + item[1] + '</option>');
```
Comment (gigorok) [inline app/assets/javascripts/ajax-chosen.js:71]: select.append('<option value="' + item.id + '">' + item.value + '</option>');

File: app/admin/routing/routing_plan_static_route.rb [hunk 10]
```diff
@@ -30,7 +30,7 @@
   filter :prefix
   filter :country, input_html: { class: 'chosen' }
   filter :network, input_html: { class: 'chosen' }
-  filter :vendor, collection: -> { Contractor.vendors }, input_html: { class: 'chosen' }
+  filter_ajax :vendor, class_name: 'Contractor', url: '/contractors/search?q[vendor_eq]=true'
```
Comment (gigorok) [inline app/admin/routing/routing_plan_static_route.rb:33]: filter :vendor, as: :select, input_html: { class: 'chosen chosen-ajax', data: { class_name: 'Contractor', path: search_contractors_path(q: {vendor: true}) } }

File: config/environments/production.rb [hunk 11]
```diff
@@ -25,7 +25,8 @@
   config.public_file_server.enabled = ENV['RAILS_SERVE_STATIC_FILES'].present?
 
   # Compress JavaScripts and CSS.
-  config.assets.js_compressor = :uglifier
+  # https://github.com/lautis/uglifier/issues/127#issuecomment-352224986
+  config.assets.js_compressor = Uglifier.new(harmony: true)
```
Comment (gigorok) [inline config/environments/production.rb:29]: upgrade uglifier instead, in separated commit

File: app/models/contractor.rb [hunk 12]
```diff
@@ -30,6 +30,7 @@ class Contractor < ActiveRecord::Base
 
   scope :customers, -> { where customer: true }
   scope :vendors, -> { where vendor: true }
+  scope :search_for, ->(term) { where('name LIKE ?', "%#{term}%") }
```
Comment (gigorok) [inline app/models/contractor.rb:33]: scope :search_for, ->(term) { where('CONCAT_WS(name, ' | ', id::varchar) LIKE ?', "%#{term}%") }

File: app/models/account.rb [hunk 13]
```diff
@@ -58,6 +58,8 @@ class Account < Yeti::ActiveRecord
   scope :vendors_accounts, -> { joins(:contractor).where('contractors.vendor' => true) }
   scope :customers_accounts, -> { joins(:contractor).where('contractors.customer' => true) }
   scope :collection, -> { order(:name) }
+  scope :search_for, ->(term) { where('name LIKE ?', "%#{term}%") }
```
Comment (gigorok) [inline app/models/account.rb:61]: follow display_name method here

File: app/assets/javascripts/active_admin.js [hunk 14]
```diff
@@ -38,13 +38,15 @@
...
 //= require import_apply_unique_fields
+//= require ajax-chosen
 
 
 $(document).ready(function () {
     $("select.chosen").chosen({no_results_text: "No results matched", width: '240px', search_contains: true});
     $("select.chosen-wide").chosen({no_results_text: "No results matched", width: '80%', search_contains: true});
     $("select.chosen-sortable").chosen({no_results_text: "No results matched", width: '80%', search_contains: true}).chosenSortable();
 
+    $("select.chosen-ajax").chosen_ajax({allow_single_deselect: true, ajax_method: "GET", ajax_min_chars: 1, no_results_text: "No results matched"})
```
Comment (gigorok) [inline app/assets/javascripts/active_admin.js:49]: ajax_min_chars better to increase to 3

File: spec/features/ajax_filters_spec.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      # make first element selected
+      find('#q_customer_id_chosen li', text: customer_name).click
+
+      # execute filter and reload page
+      page.find('input[type=submit]').click
+
+      expect(page.current_url).to include('customer_id_eq')
+      # check selected element
+      result = find('#sidebar #new_q #q_customer_input select option', visible: :hidden, text: customer_name).selected?
+      expect(result).to eq true
```
Comment (gigorok) [inline spec/features/ajax_filters_spec.rb:51]: expect(find('#sidebar #new_q #q_customer_input select option', visible: :hidden, text: customer_name)).to be_selected

File: spec/features/ajax_filters_spec.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    it 'should select option and execute filter ' do
+      customer_name = goal_customer_list.first.name
+      load_options
+      # make first element selected
+      find('#q_customer_id_chosen li', text: customer_name).click
+
+      # execute filter and reload page
+      page.find('input[type=submit]').click
+
+      expect(page.current_url).to include('customer_id_eq')
```
Comment (gigorok) [inline spec/features/ajax_filters_spec.rb:48]: expect(page.current_url).to include("q[customer_id_eq]=#{customer.id}")

File: spec/features/ajax_filters_spec.rb [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+  context 'customer filter' do
+    let!(:goal_customer_list) {
+      [
+        FactoryBot.create(:customer, name: 'customer_1'),
+        FactoryBot.create(:customer, name: 'customer_2')
+      ]
+    }
+    let!(:other_customer) { FactoryBot.create(:customer, name: 'other') }
+
+    subject(:load_options) do |string = nil|
```
Comment (gigorok) [inline spec/features/ajax_filters_spec.rb:17]: does not pass arguments to subject

File: spec/features/ajax_filters_spec.rb (hunk 16 above)
Comment (Ivanov-Anton) [reply spec/features/ajax_filters_spec.rb:48]:
```suggestion
      expect(CGI.unescape(page.current_url)).to include('customer_id_eq')
```

**Pull request merged**